src/redisearch_rs/inverted_index/src/lib.rs

```diff
@@ -253,58 +274,109 @@ pub struct RSIndexResult {
+            field_mask: RS_FIELDMASK_ALL,
+            freq: 1,
             data: RSIndexResultData {
                 num: ManuallyDrop::new(RSNumericRecord(num)),
             },
             result_type: RSResultType::Numeric,
-            is_copy: false,
-            metrics: std::ptr::null_mut(),
-            weight: 0.0,
+            weight: 1.0,
+            ..Default::default()
         }
     }
 
-    /// Create a new virtual index result
-    pub fn virt(doc_id: t_docId, field_mask: FieldMask, weight: f64) -> Self {
+    /// Create a new metric index result with the given document ID
+    pub fn metric(doc_id: t_docId) -> Self {
         Self {
             doc_id,
-            dmd: std::ptr::null(),
-            field_mask,
-            freq: 0,
-            offsets_sz: 0,
+            field_mask: RS_FIELDMASK_ALL,
             data: RSIndexResultData {
-                virt: ManuallyDrop::new(RSVirtualResult),
+                num: ManuallyDrop::new(RSNumericRecord(0.0)),
             },
-            result_type: RSResultType::Virtual,
-            is_copy: false,
-            metrics: std::ptr::null_mut(),
+            result_type: RSResultType::Metric,
+            weight: 1.0,
+            ..Default::default()
+        }
+    }
+
+    /// Create a new intersection index result with the given document ID, capacity, and weight
+    pub fn intersect(doc_id: t_docId, cap: usize, weight: f64) -> Self {
+        Self {
+            doc_id,
+            data: RSIndexResultData {
+                agg: ManuallyDrop::new(RSAggregateResult::with_capacity(cap)),
+            },
+            result_type: RSResultType::Intersection,
             weight,
+            ..Default::default()
         }
     }
 
     /// Create a new union index result with the given document ID, capacity, and weight
     pub fn union(doc_id: t_docId, cap: usize, weight: f64) -> Self {
         Self {
             doc_id,
-            dmd: std::ptr::null(),
-            field_mask: 0,
-            freq: 0,
-            offsets_sz: 0,
             data: RSIndexResultData {
                 agg: ManuallyDrop::new(RSAggregateResult::with_capacity(cap)),
             },
             result_type: RSResultType::Union,
-            is_copy: false,
-            metrics: std::ptr::null_mut(),
             weight,
+            ..Default::default()
+        }
+    }
+
+    /// Create a new hybrid metric index result with the given document ID
+    pub fn hybrid_metric(doc_id: t_docId) -> Self {
+        Self {
+            doc_id,
+            data: RSIndexResultData {
+                agg: ManuallyDrop::new(RSAggregateResult::with_capacity(2)),
+            },
+            result_type: RSResultType::HybridMetric,
+            weight: 1.0,
+            ..Default::default()
+        }
+    }
+
+    /// Create a new term index result with the given document ID, term pointer, and weight
+    pub fn term(doc_id: t_docId, term: *mut RSQueryTerm, weight: f64) -> Self {
```

Comment:
My version is https://github.com/RediSearch/RediSearch/pull/6493/commits/c14d8b73f9e0113424c480d527446e4a8c5bb95c

I guess we should agree on what should be passed to those constructors and what's should be default.

Reply:
Yes, let's put our heads together on it. We should be able to construct all the default ones the same as this C file, but still have an easy way to make ones for tests.

https://github.com/RediSearch/RediSearch/blob/0c8a28b0ad094119ce32872542b49da9e293a057/src/index_result.c

Small details that are important are things like setting the fieldmask for numeric records correctly (to ALL). I only found the Rust constructor (and therefore decoder) was wrong when I compared the inverted index output to the C in a benchmark.

Reply:
> We should be able to construct all the default ones the same as this C file, but still have an easy way to make ones for tests.

The builder pattern is the right solution here then.

Reply:
Builder done in https://github.com/RediSearch/RediSearch/pull/6495/commits/b75aa0be5393e20237c57b706bbd022e4fe1189b

I'm not sure if we want to make a proper builder struct yet though. It is possible, but I worry it might make the code too complex.